libraries/core_libs/consensus/src/dag/dag.cpp

```diff
@@ -422,20 +392,26 @@ bool DagManager::addDagBlock(DagBlock &&blk, SharedTransactions &&trxs, bool pro
     {
       ULock lock(mutex_);
       if (save) {
-        if (db_->dagBlockInDb(blk_hash)) {
-          LOG(log_dg_) << "Block already in DB: " << blk_hash;
-          return false;
+        if (db_->dagBlockInDb(blk.getHash())) {
+          // It is a valid scenario that two threads can receive same block from two peers and process at same time
+          return {true, std::nullopt};
         }
-        const auto dag_expiry_level = dag_blk_mgr_->getDagExpiryLevel();
-        if (blk.getLevel() < dag_expiry_level) {
-          LOG(log_nf_) << "Dropping old block: " << blk_hash << ". Expiry level: " << dag_expiry_level
+
+        if (blk.getLevel() < dag_expiry_level_) {
+          LOG(log_nf_) << "Dropping old block: " << blk_hash << ". Expiry level: " << dag_expiry_level_
                        << ". Block level: " << blk.getLevel();
-          return false;
+          return {false, std::nullopt};
+        }
+
+        auto res = pivotAndTipsAvailable(blk);
```

Comment:
Why is it changed to return a pair, but second is not used?

Reply:
Ok, I see that we use it to add to disconnect reason for peer later. But would it be useful? This pair looks pretty weird for me here and we aren't sending all missing hashes, but only first one we found(we could miss a few, right?). So it is not very useful as I see

Reply:
Maybe it would be cleaner to change this method to `missingPivotOrTips` and return collection with missing hashes. If it is empty we know that we have all of them

Reply:
Done